instrumentation/aws-sdk/aws-sdk-1.11/javaagent/src/test/groovy/S3TracingTest.groovy

```diff
@@ -0,0 +1,449 @@
+      .withEnv("DEBUG", "1")
+      .withEnv("SQS_PROVIDER", "elasticmq")
+    localstack.start()
+
+    sqsClient = AmazonSQSAsyncClient.asyncBuilder()
+      .withEndpointConfiguration(localstack.getEndpointConfiguration(LocalStackContainer.Service.SQS))
+      .withCredentials(localstack.getDefaultCredentialsProvider())
+      //.withRegion(Regions.US_EAST_1)
+      .build()
+
+    s3Client = AmazonS3Client.builder()
+      .withEndpointConfiguration(localstack.getEndpointConfiguration(LocalStackContainer.Service.SNS))
+      .withCredentials(localstack.getDefaultCredentialsProvider())
+      //.withRegion(Regions.US_EAST_1)
+      .build()
+
+    localstack.followOutput(new Slf4jLogConsumer(LoggerFactory.getLogger("test")))
+  }
+
+  def cleanupSpec() {
+    if (localstack != null) {
+      localstack.stop()
+    }
+  }
+
+  def createQueue(String queueName) {
+    return sqsClient.createQueue(queueName).getQueueUrl()
+  }
+
+  def getQueueArn(String queueUrl) {
+    return sqsClient.getQueueAttributes(
+      new GetQueueAttributesRequest(queueUrl)
+        .withAttributeNames("QueueArn")).getAttributes()
+      .get("QueueArn")
+  }
+
+  def setQueuePolicy(String queueUrl, String queueArn) {
+    sqsClient.setQueueAttributes(queueUrl, Collections.singletonMap("Policy", policy(queueArn)))
+  }
+
+  def createBucket(String bucketName) {
+    s3Client.createBucket(bucketName)
+  }
+
+  def deleteBucket(String bucketName) {
+    ObjectListing objectListing = s3Client.listObjects(bucketName)
+      Iterator<S3ObjectSummary> objIter = objectListing.getObjectSummaries().iterator()
+      while (objIter.hasNext()) {
+        s3Client.deleteObject(bucketName, objIter.next().getKey())
+      }
+    s3Client.deleteBucket(bucketName)
+  }
+
+  def enableS3Notifications(String bucketName, String sqsQueueArn) {
+    BucketNotificationConfiguration notificationConfiguration = new BucketNotificationConfiguration()
+    notificationConfiguration.addConfiguration("sqsQueueConfig",
+      new QueueConfiguration(sqsQueueArn, EnumSet.of(S3Event.ObjectCreatedByPut)))
+    s3Client.setBucketNotificationConfiguration(new SetBucketNotificationConfigurationRequest(
+      bucketName, notificationConfiguration))
+  }
+
+  @Ignore("Requires https://github.com/localstack/localstack/issues/3686 to work with localstack")
+  def "simple S3 upload as producer - SQS consumer services"() {
+    setup:
+    String queueName = "s3ToSqsTestQueue"
+    String bucketName = "s3-sqs-test-bucket"
+
+    String queueUrl = createQueue(queueName)
+    String queueArn = getQueueArn(queueUrl)
+    setQueuePolicy(queueUrl, queueArn)
+    createBucket(bucketName)
+    enableS3Notifications(bucketName, queueArn)
+
+    when:
+    Thread.sleep(2000)
+    // test message, auto created by AWS
+    sqsClient.receiveMessage(new ReceiveMessageRequest(queueUrl))
+    s3Client.putObject(bucketName, "testKey", "testData")
+    Thread.sleep(3000)
```

Comment:
Is it possible to replace `Thread.sleep` with intermediate `assertTraces` ?

Reply:
Or at least `waitForTraces`

Reply:
I don't think that's possible. In both cases sleep() is added to give the infrastructure (AWS now, localstack in the future) some time to propagate requested sync.
First sleep - enabling S3 notifications to SQS queue and sending "test" message (consumed in the next step). If we don't do that, receive message will still execute but won't consume the message (and will break the asserts).
Second sleep - gives S3 time to create notification message and deliver it to SQS. 

I believe that once appropriate localstack feature is implemented, the test will run *without* any sleeps (execution of notification will be sync with s3 upload).

@mateuszrzeszutek & @iNikem ^^

Reply:
> First sleep - enabling S3 notifications to SQS queue and sending "test" message

Does this result in a trace? If yes, then we can wait for this trace to arrive instead of sleep, no?

Reply:
The problem with SQS consumer is that it will create a similar trace for "ReceiveMessage" with and without consumed message. So if we send "Receive" and wait for trace we will get one, but "test" message may not be consumed and break the trace later on (we check for parent-child link and for "test" message we won't get one). 

To avoid explicit "Thread.sleep()" we could utilise "getWaitTimeSeconds" property of the ReceiveMessageRequest - albeit it will still mean waiting ;) @iNikem LMK if this would be better.

Reply:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-short-and-long-polling.html#sqs-long-polling

This reads to me that you can tell `receiveMessage` to wait until message is available. Thus we will wait only as long as needed and not defensively for 2-3 seconds in any case.

Reply:
Yeah, that's what I meant (same setting). Will test with AWS and commit change if it works as expected :)